Install features attempts to install previously generated features

To reproduce, use a project similar to the `basic-dev-project` in `dev-it` that requires `jaxrs` and `servlet` features, but does not specify any umbrella dependencies in the POM.
Running dev mode to start with `mvn liberty:dev` should generate `jaxrs-2.1` and `servlet-4.0` in the `generated-features.xml`, as shown in the output below:
```
[INFO] Generated the following features: [servlet-4.0, batch-1.0, jaxrs-2.1]
```
Then, add `webProfile-7.0` to `server.xml` or any other server configuration file. The following are snippets from the dev mode console:
```
[INFO] Generating optimized features list...
[INFO] Running liberty:generate-features
[INFO] Parsing the server file C:\Users\TREVORCrawford\Git\modified-demo-devmode\src\main\liberty\config\server.xml for features and includes.
[INFO] Parsing the server file C:\Users\TREVORCrawford\Git\modified-demo-devmode\src\main\liberty\config\extraFeatures.xml for features and includes.
[INFO] Features were included for file extraFeatures.xml
[INFO] Generated the following features: [servlet-3.1, batch-1.0, jaxrs-2.0]
...
[INFO] Configuration features have been added
[INFO] Running liberty:install-feature
...
[INFO] Installing features: [webprofile-7.0, servlet-4.0, batch-1.0, jaxrs-2.1]

[ERROR] Failed to install features from configuration file
org.apache.maven.plugin.MojoExecutionException: A feature conflict error occurred while installing features: [webprofile-7.0, servlet-4.0, batch-1.0, jaxrs-2.1] ...
```

Binary scanner generates lower versions of the `jaxrs` and `servlet` features to be compatible with `webProfile-7.0`, which is the correct behaviour if no umbrella dependency is specified. However, it appears the `install-feature` goal is retrieving an old version of the generated features list and runs into a feature conflict error.

This issue appears to sort itself out the next time generate features is run (if the feature list has not been modified):
```
[INFO] Generated the following features: [servlet-3.1, batch-1.0, jaxrs-2.0]
...
[INFO] Configuration features have been added
[INFO] Running liberty:install-feature
...
[INFO] Installing features: [webprofile-7.0, servlet-3.1, batch-1.0, jaxrs-2.0]
```

The root of this issue seems to be the case where a feature is removed OR version is changed (dev mode views as a feature removal and feature addition) in ANY configuration xml file (does not apply to just the generate-features case, although we are seeing this bug more now because of the continuous changes to the generated-features.xml file). 

To reproduce
1. Start dev mode `mvn liberty:dev -DgenerateFeatures=false` with `servlet-3.1` in the server.xml
2. Modify the server.xml so that `servlet-3.1` becomes `servlet-4.0`
3. Install feature fails due to conflict between `servlet-3.1` and `servlet-4.0`:
```xml
[INFO] Copied file: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/src/main/liberty/config/server.xml to: /private/var/folders/gr/3p093xzj0f9fj9psx80z_zw00000gn/T/tempConfig14803358266236911421/server.xml
[INFO] Parsing the server file /private/var/folders/gr/3p093xzj0f9fj9psx80z_zw00000gn/T/tempConfig14803358266236911421/server.xml for features and includes.
[INFO] Parsing the server file /private/var/folders/gr/3p093xzj0f9fj9psx80z_zw00000gn/T/tempConfig14803358266236911421/extraFeatures.xml for features and includes.
[INFO] Parsing the server file /private/var/folders/gr/3p093xzj0f9fj9psx80z_zw00000gn/T/tempConfig14803358266236911421/configDropins/overrides/generated-features.xml for features and includes.
[INFO] Configuration features have been removed
[INFO] Configuration features have been added
[INFO] Running liberty:install-feature
[INFO] CWWKM2102I: Using artifact based assembly archive : io.openliberty:openliberty-kernel:null:22.0.0.3:zip.
[INFO] CWWKM2102I: Using installDirectory : /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp.
[INFO] CWWKM2102I: Using serverName : defaultServer.
[INFO] CWWKM2102I: Using serverDirectory : /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer.
[INFO] Parsing the server file /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/server.xml for features and includes.
[INFO] Parsing the server file /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/extraFeatures.xml for features and includes.
[INFO] Parsing the server file /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/configDropins/overrides/generated-features.xml for features and includes.
[INFO] Installing features: [servlet-4.0, servlet-3.1]

[ERROR] Failed to install features from configuration file
org.apache.maven.plugin.MojoExecutionException: A feature conflict error occurred while installing features: [servlet-4.0, servlet-3.1]: CWWKF1405E: The servlet-4.0 and servlet-3.1 singleton features are incompatible with each other and they cannot be installed at the same time.

    at org.codehaus.mojo.pluginsupport.MojoSupport.execute (MojoSupport.java:137)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:120)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojo (StartDebugMojoSupport.java:354)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojoInstallFeature (StartDebugMojoSupport.java:339)
    at io.openliberty.tools.maven.server.DevMojo.runLibertyMojoInstallFeature (DevMojo.java:1843)
    at io.openliberty.tools.maven.server.DevMojo$DevMojoUtil.installFeatures (DevMojo.java:991)
    at io.openliberty.tools.common.plugins.util.DevUtil.installFeaturesToTempDir (DevUtil.java:4397)
    at io.openliberty.tools.common.plugins.util.DevUtil.processFileChanges (DevUtil.java:4079)
    at io.openliberty.tools.common.plugins.util.DevUtil.access$1500 (DevUtil.java:97)
    at io.openliberty.tools.common.plugins.util.DevUtil$8.onAlteration (DevUtil.java:3284)
    at io.openliberty.tools.common.plugins.util.DevUtil$8.onFileChange (DevUtil.java:3255)
    at org.apache.commons.io.monitor.FileAlterationObserver.doMatch (FileAlterationObserver.java:412)
    at org.apache.commons.io.monitor.FileAlterationObserver.checkAndNotify (FileAlterationObserver.java:335)
    at org.apache.commons.io.monitor.FileAlterationObserver.checkAndNotify (FileAlterationObserver.java:305)
    at io.openliberty.tools.common.plugins.util.DevUtil.watchFiles (DevUtil.java:3076)
    at io.openliberty.tools.maven.server.DevMojo.doExecute (DevMojo.java:1372)
    at org.codehaus.mojo.pluginsupport.MojoSupport.execute (MojoSupport.java:122)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: io.openliberty.tools.common.plugins.util.PluginExecutionException: A feature conflict error occurred while installing features: [servlet-4.0, servlet-3.1]: CWWKF1405E: The servlet-4.0 and servlet-3.1 singleton features are incompatible with each other and they cannot be installed at the same time.

    at io.openliberty.tools.common.plugins.util.InstallFeatureUtil.installFeatures (InstallFeatureUtil.java:607)
    at io.openliberty.tools.maven.server.InstallFeatureMojo.installFeatures (InstallFeatureMojo.java:87)
    at io.openliberty.tools.maven.server.InstallFeatureMojo.doExecute (InstallFeatureMojo.java:58)
    at org.codehaus.mojo.pluginsupport.MojoSupport.execute (MojoSupport.java:122)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:120)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojo (StartDebugMojoSupport.java:354)
    at io.openliberty.tools.maven.server.StartDebugMojoSupport.runLibertyMojoInstallFeature (StartDebugMojoSupport.java:339)
    at io.openliberty.tools.maven.server.DevMojo.runLibertyMojoInstallFeature (DevMojo.java:1843)
    at io.openliberty.tools.maven.server.DevMojo$DevMojoUtil.installFeatures (DevMojo.java:991)
    at io.openliberty.tools.common.plugins.util.DevUtil.installFeaturesToTempDir (DevUtil.java:4397)
    at io.openliberty.tools.common.plugins.util.DevUtil.processFileChanges (DevUtil.java:4079)
    at io.openliberty.tools.common.plugins.util.DevUtil.access$1500 (DevUtil.java:97)
    at io.openliberty.tools.common.plugins.util.DevUtil$8.onAlteration (DevUtil.java:3284)
    at io.openliberty.tools.common.plugins.util.DevUtil$8.onFileChange (DevUtil.java:3255)
    at org.apache.commons.io.monitor.FileAlterationObserver.doMatch (FileAlterationObserver.java:412)
    at org.apache.commons.io.monitor.FileAlterationObserver.checkAndNotify (FileAlterationObserver.java:335)
    at org.apache.commons.io.monitor.FileAlterationObserver.checkAndNotify (FileAlterationObserver.java:305)
    at io.openliberty.tools.common.plugins.util.DevUtil.watchFiles (DevUtil.java:3076)
    at io.openliberty.tools.maven.server.DevMojo.doExecute (DevMojo.java:1372)
    at org.codehaus.mojo.pluginsupport.MojoSupport.execute (MojoSupport.java:122)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[INFO] Copied file: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/src/main/liberty/config/server.xml to: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/server.xml
[INFO] [AUDIT   ] CWWKG0016I: Starting server configuration update.
[INFO] [AUDIT   ] CWWKG0028A: Processing included configuration resource: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/extraFeatures.xml
[INFO] [AUDIT   ] CWWKG0093A: Processing configuration drop-ins resource: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/configDropins/overrides/generated-features.xml
[INFO] [ERROR   ] CWWKF0001E: A feature definition could not be found for servlet-4.0
[INFO] [AUDIT   ] CWWKT0017I: Web application removed (default_host): http://192.168.0.32:9080/dev-sample-proj-1.0-SNAPSHOT/
[INFO] [AUDIT   ] CWWKZ0009I: The application dev-sample-proj-1.0-SNAPSHOT has stopped successfully.
[INFO] [AUDIT   ] CWWKG0017I: The server configuration was successfully updated in 0.486 seconds.
[INFO] [AUDIT   ] CWWKF0013I: The server removed the following features: [servlet-3.1].
[INFO] [AUDIT   ] CWWKF0008I: Feature update completed in 0.473 seconds

```
4. Save the server.xml file again, all works as expected:
```xml
[INFO] Copied file: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/src/main/liberty/config/server.xml to: /private/var/folders/gr/3p093xzj0f9fj9psx80z_zw00000gn/T/tempConfig6685235736575724630/server.xml
[INFO] Parsing the server file /private/var/folders/gr/3p093xzj0f9fj9psx80z_zw00000gn/T/tempConfig6685235736575724630/server.xml for features and includes.
[INFO] Parsing the server file /private/var/folders/gr/3p093xzj0f9fj9psx80z_zw00000gn/T/tempConfig6685235736575724630/extraFeatures.xml for features and includes.
[INFO] Parsing the server file /private/var/folders/gr/3p093xzj0f9fj9psx80z_zw00000gn/T/tempConfig6685235736575724630/configDropins/overrides/generated-features.xml for features and includes.
[INFO] Configuration features have been added
[INFO] Running liberty:install-feature
[INFO] CWWKM2102I: Using artifact based assembly archive : io.openliberty:openliberty-kernel:null:22.0.0.3:zip.
[INFO] CWWKM2102I: Using installDirectory : /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp.
[INFO] CWWKM2102I: Using serverName : defaultServer.
[INFO] CWWKM2102I: Using serverDirectory : /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer.
[INFO] Parsing the server file /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/server.xml for features and includes.
[INFO] Parsing the server file /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/extraFeatures.xml for features and includes.
[INFO] Parsing the server file /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/configDropins/overrides/generated-features.xml for features and includes.
[INFO] Installing features: [servlet-4.0]

[INFO] Product validation completed successfully.
[INFO] The following features have been installed: servlet-4.0 
[INFO] Copied file: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/src/main/liberty/config/server.xml to: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/server.xml
[INFO] [AUDIT   ] CWWKG0016I: Starting server configuration update.
[INFO] [AUDIT   ] CWWKG0028A: Processing included configuration resource: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/extraFeatures.xml
[INFO] [AUDIT   ] CWWKG0093A: Processing configuration drop-ins resource: /Users/kathrynkodama/devex/featureGeneration/basic-dev-project/target/liberty/wlp/usr/servers/defaultServer/configDropins/overrides/generated-features.xml
[INFO] [AUDIT   ] CWWKG0018I: The server configuration was not updated. No functional changes were detected.

```

Dev mode is running. Add a Java file:
```
/private/tmp/basic-dev/src/main/java/com/demo/SystemHealth.java:4: error: package org.eclipse.microprofile.health does not exist
import org.eclipse.microprofile.health.Health;
                                      ^
etc...
```
Add a dependency to pom.xml. As soon as Save is executed this runs:
```
[INFO] Running liberty:generate-features
...
[INFO] Running liberty:deploy
...
```
Next recompile is attempted and config files are copied. Liberty updates server configuration.

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/base-install-feature-test/src/test/java/net/wasdev/wlp/test/feature/it/BaseInstallFeature.java
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/base-install-feature-test/src/test/java/net/wasdev/wlp/test/feature/it/BaseInstallFeature.java
@@ -19,7 +19,7 @@ import java.io.File;
 import java.io.FilenameFilter;
 import io.openliberty.tools.common.plugins.util.InstallFeatureUtil;
 
-import static junit.framework.Assert.*;
+import static org.junit.Assert.assertEquals;
 import org.junit.Before;
 
 public class BaseInstallFeature {

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/pom.xml
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/pom.xml
@@ -1,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.openliberty.tools.it</groupId>
+        <artifactId>kernel-install-feature-tests</artifactId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    
+    <artifactId>install-features-override-server-dir-it</artifactId>
+    <packaging>jar</packaging>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>io.openliberty.tools</groupId>
+                <artifactId>liberty-maven-plugin</artifactId>
+                <version>@pom.version@</version>
+                <configuration>
+                    <assemblyArtifact>
+                        <groupId>${runtimeGroupId}</groupId>
+                        <artifactId>${runtimeKernelId}</artifactId>
+                        <version>${runtimeVersion}</version>
+                        <type>zip</type>
+                    </assemblyArtifact>
+                    <serverName>test</serverName>
+                    <serverDir>src/test/resources</serverDir>
+                </configuration>
+                <executions>
+                    <execution>
+                        <id>install-liberty-server</id>
+                        <phase>compile</phase>
+                        <goals>
+                            <goal>install-server</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>create-server</id>
+                        <phase>pre-integration-test</phase>
+                        <goals>
+                            <goal>create</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>install-server-features</id>
+                        <phase>pre-integration-test</phase>
+                        <goals>
+                            <goal>install-feature</goal>
+                        </goals>
+                        <configuration>
+                            <features>
+                                <acceptLicense>true</acceptLicense>
+                            </features>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+</project>

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/pom.xml
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/pom.xml
@@ -26,7 +26,6 @@
                         <type>zip</type>
                     </assemblyArtifact>
                     <serverName>test</serverName>
-                    <serverDir>src/test/resources</serverDir>
                 </configuration>
                 <executions>
                     <execution>
@@ -53,6 +52,7 @@
                             <features>
                                 <acceptLicense>true</acceptLicense>
                             </features>
+                            <serverDir>src/test/resources</serverDir>
                         </configuration>
                     </execution>
                 </executions>

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/java/net/wasdev/wlp/test/features/it/InstallFeaturesOverrideServerDirTest.java
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/java/net/wasdev/wlp/test/features/it/InstallFeaturesOverrideServerDirTest.java
@@ -22,8 +22,10 @@ public class InstallFeaturesOverrideServerDirTest extends BaseInstallFeature {
     @Test
     public void testInstalledFeatures() throws Exception {
         assertInstalled("jsp-2.3"); // specified in extra-features.xml
-        assertNotInstalled("servlet-3.1"); // specified in the server.xml
+        // ensure that the serverDir param is working as expected and 
+        // servlet-3.1 is not installed as servlet-4.0 is specified
+        assertNotInstalled("servlet-3.1");
         assertInstalled("appSecurityClient-1.0"); // specified in the server.xml
-        assertInstalled("servlet-4.0"); // not specified in server.xml
+        assertInstalled("servlet-4.0"); // specified in server.xml
     }
 }

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/java/net/wasdev/wlp/test/features/it/InstallFeaturesOverrideServerDirTest.java
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/java/net/wasdev/wlp/test/features/it/InstallFeaturesOverrideServerDirTest.java
@@ -1,0 +1,28 @@
+/*******************************************************************************
+ * (c) Copyright IBM Corporation 2022.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package net.wasdev.wlp.test.feature.it;
+
+import org.junit.Test;
+
+public class InstallFeaturesOverrideServerDirTest extends BaseInstallFeature {
+    
+    @Test
+    public void testInstalledFeatures() throws Exception {
+        assertNotInstalled("servlet-4.0");
+        assertInstalled("jsp-2.3");
+        assertInstalled("servlet-3.1");
+    }
+}

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/java/net/wasdev/wlp/test/features/it/InstallFeaturesOverrideServerDirTest.java
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/java/net/wasdev/wlp/test/features/it/InstallFeaturesOverrideServerDirTest.java
@@ -21,8 +21,9 @@ public class InstallFeaturesOverrideServerDirTest extends BaseInstallFeature {
     
     @Test
     public void testInstalledFeatures() throws Exception {
-        assertNotInstalled("servlet-4.0");
-        assertInstalled("jsp-2.3");
-        assertInstalled("servlet-3.1");
+        assertInstalled("jsp-2.3"); // specified in extra-features.xml
+        assertNotInstalled("servlet-3.1"); // specified in the server.xml
+        assertInstalled("appSecurityClient-1.0"); // specified in the server.xml
+        assertInstalled("servlet-4.0"); // not specified in server.xml
     }
 }

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/resources/extraFeatures.xml
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/resources/extraFeatures.xml
@@ -1,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<server description="extra server features">
+	<featureManager>
+        <feature>jsp-2.3</feature>
+    </featureManager>
+</server>

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/resources/server.xml
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/resources/server.xml
@@ -1,0 +1,7 @@
+<server description="default server">
+    <featureManager>    
+        <feature>servlet-4.0</feature>
+    </featureManager> 
+
+    <include location="extraFeatures.xml"/>    
+</server>

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/resources/server.xml
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/install-features-override-server-dir-it/src/test/resources/server.xml
@@ -1,6 +1,7 @@
 <server description="default server">
     <featureManager>    
         <feature>servlet-4.0</feature>
+        <feature>appSecurityClient-1.0</feature>
     </featureManager> 
 
     <include location="extraFeatures.xml"/>    

--- a/liberty-maven-plugin/src/it/kernel-install-feature-test/pom.xml
+++ b/liberty-maven-plugin/src/it/kernel-install-feature-test/pom.xml
@@ -40,6 +40,7 @@
         <module>install-ol-features-no-accept-license-it</module> 
         <module>prepare-feature-it</module> 
         <module>install-usr-feature-ext-it</module>
+        <module>install-features-override-server-dir-it</module>
     </modules>
     
     <!-- Profiles for WLP vs OL -->

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java
@@ -397,7 +397,7 @@ public class DevMojo extends LooseAppSupport {
         @Override
         public void libertyInstallFeature() throws PluginExecutionException {
             try {
-                runLibertyMojoInstallFeature(null, container ? super.getContainerName() : null);
+                runLibertyMojoInstallFeature(null, null, container ? super.getContainerName() : null);
             } catch (MojoExecutionException e) {
                 throw new PluginExecutionException(e);
             }
@@ -941,7 +941,7 @@ public class DevMojo extends LooseAppSupport {
                     	runLibertyMojoDeploy();
                     }
                     if (installFeature) {
-                        runLibertyMojoInstallFeature(null, super.getContainerName());
+                        runLibertyMojoInstallFeature(null, null, super.getContainerName());
                     }
                 }
                 if (!(restartServer || createServer || redeployApp || installFeature || runBoostPackage)) {
@@ -974,22 +974,24 @@ public class DevMojo extends LooseAppSupport {
                         Set<String> existingFeaturesCopy = new HashSet<String>(existingFeatures);
                         existingFeaturesCopy.removeAll(featuresCopy);
                         if (!existingFeaturesCopy.isEmpty()) {
-                            log.info("Configuration features have been removed");
+                            log.info("Configuration features have been removed: " + existingFeaturesCopy);
                             existingFeatures.removeAll(existingFeaturesCopy);
                         }
                     }
 
                     // check if features have been added and install new features
                     if (!features.isEmpty()) {
-                        log.info("Configuration features have been added");
+                        log.info("Configuration features have been added: " + features);
+                        // pass all features to install feature
+                        existingFeatures.addAll(features);
+
                         Element[] featureElems = new Element[features.size() + 1];
                         featureElems[0] = element(name("acceptLicense"), "true");
                         String[] values = features.toArray(new String[features.size()]);
                         for (int i = 0; i < features.size(); i++) {
                             featureElems[i + 1] = element(name("feature"), values[i]);
                         }
-                        runLibertyMojoInstallFeature(element(name("features"), featureElems), super.getContainerName());
-                        existingFeatures.addAll(features);
+                        runLibertyMojoInstallFeature(element(name("features"), featureElems), serverDir, super.getContainerName());
                     }
                 }
             } catch (MojoExecutionException e) {
@@ -1265,7 +1267,7 @@ public class DevMojo extends LooseAppSupport {
             // should have "RUN features.sh" in their Dockerfile if they want features to be
             // installed.
             if (!container) {
-                runLibertyMojoInstallFeature(null, null);
+                runLibertyMojoInstallFeature(null, null, null);
             }
             runLibertyMojoDeploy();
         }
@@ -1840,8 +1842,8 @@ public class DevMojo extends LooseAppSupport {
      * @throws MojoExecutionException
      */
     @Override
-    protected void runLibertyMojoInstallFeature(Element features, String containerName) throws MojoExecutionException {
-        super.runLibertyMojoInstallFeature(features, containerName);
+    protected void runLibertyMojoInstallFeature(Element features, File serverDir, String containerName) throws MojoExecutionException {
+        super.runLibertyMojoInstallFeature(features, serverDir, containerName);
     }
 
     /**

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java
@@ -982,9 +982,8 @@ public class DevMojo extends LooseAppSupport {
                     // check if features have been added and install new features
                     if (!features.isEmpty()) {
                         log.info("Configuration features have been added: " + features);
-                        // pass all features to install feature
                         existingFeatures.addAll(features);
-
+                        // pass all new features to install-feature
                         Element[] featureElems = new Element[features.size() + 1];
                         featureElems[0] = element(name("acceptLicense"), "true");
                         String[] values = features.toArray(new String[features.size()]);

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java
@@ -983,7 +983,7 @@ public class DevMojo extends LooseAppSupport {
                     if (!features.isEmpty()) {
                         log.info("Configuration features have been added: " + features);
                         existingFeatures.addAll(features);
-                        // pass all new features to install-feature
+                        // pass all new features to install-feature as backup in case the serverDir cannot be accessed
                         Element[] featureElems = new Element[features.size() + 1];
                         featureElems[0] = element(name("acceptLicense"), "true");
                         String[] values = features.toArray(new String[features.size()]);

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/InstallFeatureMojo.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/InstallFeatureMojo.java
@@ -51,8 +51,8 @@ public class InstallFeatureMojo extends InstallFeatureSupport {
      * Dev mode uses this option to provide the location of the
      * temporary serverDir it uses after a change to the server.xml
      */
-    @Parameter(property = "serverDir")
-    private File serverDirectoryParam = null;
+    @Parameter(alias = "serverDir", property = "serverDir")
+    private File serverDirectoryParam;
 
     /*
      * (non-Javadoc)

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/InstallFeatureMojo.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/InstallFeatureMojo.java
@@ -46,6 +46,14 @@ public class InstallFeatureMojo extends InstallFeatureSupport {
     @Parameter
     private String containerName;
 
+    /**
+     * (Optional) Server directory to get the features from.
+     * Dev mode uses this option to provide the location of the
+     * temporary serverDir it uses after a change to the server.xml
+     */
+    @Parameter(property = "serverDir")
+    private File serverDirectoryParam = null;
+
     /*
      * (non-Javadoc)
      * @see org.codehaus.mojo.pluginsupport.MojoSupport#doExecute()
@@ -54,6 +62,10 @@ public class InstallFeatureMojo extends InstallFeatureSupport {
     protected void doExecute() throws Exception {
         if(!initialize()) {
             return;
+        }
+        if (serverDirectoryParam != null) {
+            serverDirectory = serverDirectoryParam;
+            log.debug("Overriding the server directory with: " + serverDirectory);
         }
         installFeatures();
     }

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/RunServerMojo.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/RunServerMojo.java
@@ -121,7 +121,7 @@ public class RunServerMojo extends PluginConfigSupport {
         }
         
         runLibertyMojoCreate();
-        runLibertyMojoInstallFeature(null, null);
+        runLibertyMojoInstallFeature(null, null, null);
         runLibertyMojoDeploy(false);
 
         ServerTask serverTask = initializeJava();

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/StartDebugMojoSupport.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/StartDebugMojoSupport.java
@@ -328,7 +328,7 @@ public class StartDebugMojoSupport extends ServerFeatureSupport {
         runLibertyMojo("deploy", config);
     }
 
-    protected void runLibertyMojoInstallFeature(Element features, String containerName) throws MojoExecutionException {
+    protected void runLibertyMojoInstallFeature(Element features, File serverDir, String containerName) throws MojoExecutionException {
         Xpp3Dom config = ExecuteMojoUtil.getPluginGoalConfig(getLibertyPlugin(), "install-feature", log);
         if (features != null) {
             config = Xpp3Dom.mergeXpp3Dom(configuration(features), config);
@@ -336,7 +336,15 @@ public class StartDebugMojoSupport extends ServerFeatureSupport {
         if (containerName != null) {
             config.addChild(element(name("containerName"), containerName).toDom());
         }
-        runLibertyMojo("install-feature", config);   
+        if (serverDir != null) {
+            try {
+                config.addChild(element(name("serverDir"), serverDir.getCanonicalPath()).toDom());
+            } catch (IOException e) {
+                log.warn("Unable to pass 'serverDir' configuration parameter to liberty:install-feature: "
+                        + serverDir);
+            }
+        }
+        runLibertyMojo("install-feature", config);
     }
 
     protected void runLibertyMojoGenerateFeatures(Element classFiles, boolean optimize) throws MojoExecutionException {

